feature/networkinstance/tests/interface_assignments/interface_assignments_test.go

```diff
@@ -0,0 +1,129 @@
+)
+
+func TestMain(m *testing.M) {
+	fptest.RunTests(m)
+}
+
+// intfNIAssignment describes a set of interfaces assigned to a network instance.
+type intfNIAssignment struct {
+	// Ports is the set of ports that are to be assigned to the network instance.
+	Ports []portSpec
+	// Type is the type specified in the configuration.
+	Type telemetry.E_NetworkInstanceTypes_NETWORK_INSTANCE_TYPE
+}
+
+// portSpec describes a specific interface, subinterface
+type portSpec struct {
+	// Name is the name of the port in the testbed.
+	Name string
+	// Subintf is the index of the subinterface on the device.
+	Subintf uint32
+	// IPv4 is the IPv4 address that should be assigned to the interface.
+	IPv4 string
+	// PrefixLength is the prefix length for the IPv4 address.
+	PrefixLength uint8
+}
+
+// TestInterfaceAssignment tests the assignment of an interface explicitly to
+// a network instance.
+func TestInterfaceAssignment(t *testing.T) {
+	tests := []struct {
+		desc          string
+		inAssignments map[string]intfNIAssignment
+		wantErr       bool
+	}{{
+		desc: "explicit assignment of port1 to DEFAULT",
+		inAssignments: map[string]intfNIAssignment{
+			*deviations.DefaultNetworkInstance: {
+				Ports: []portSpec{{
+					Name:         "port1",
+					Subintf:      0,
+					IPv4:         "192.0.2.0",
+					PrefixLength: 31,
+				}},
+				Type: telemetry.NetworkInstanceTypes_NETWORK_INSTANCE_TYPE_DEFAULT_INSTANCE,
+			},
+		},
+	}, {
+		desc: "explicit assingment of port1 to non-default NI",
+		inAssignments: map[string]intfNIAssignment{
+			"BLUE": {
+				Ports: []portSpec{{
+					Name:         "port1",
+					Subintf:      0,
+					IPv4:         "192.0.2.0",
+					PrefixLength: 31,
+				}},
+				Type: telemetry.NetworkInstanceTypes_NETWORK_INSTANCE_TYPE_L3VRF,
+			},
+		},
+	}}
+
+	for _, tt := range tests {
+		t.Run(tt.desc, func(t *testing.T) {
+			dut := ondatra.DUT(t, "dut")
+
+			d := &telemetry.Device{}
+			for niName, spec := range tt.inAssignments {
+				ni := d.GetOrCreateNetworkInstance(niName)
+				ni.Type = spec.Type
+
+				for _, p := range spec.Ports {
+
+					dp := dut.Port(t, p.Name)
+
+					// Create the interface to ensure we have no missing references.
+					intf := d.GetOrCreateInterface(dp.Name())
+					intf.Type = telemetry.IETFInterfaces_InterfaceType_ethernetCsmacd
+					intf.GetOrCreateSubinterface(p.Subintf).
+						GetOrCreateIpv4().GetOrCreateAddress(p.IPv4).PrefixLength = &p.PrefixLength
+
+					// Assign the interface to a network instance.
+					i := ni.GetOrCreateInterface(p.Name)
+					i.Interface = ygot.String(dp.Name())
+					i.Subinterface = ygot.Uint32(p.Subintf)
+				}
+			}
+			fptest.LogYgot(t, "Device", dut.Config(), d)
+
+			if tt.wantErr {
+				if got := testt.ExpectFatal(t, func(t testing.TB) {
+					dut.Config().Update(t, d)
+				}); got == "" {
+					t.Fatalf("did not get expected Fatal error, got: %s, wantErr? %v", got, tt.wantErr)
+				}
+				return
+			}
+
+			dut.Config().Update(t, d)
```

Comment:
There seems to be an order dependence where the passage of the test relies on the exact order the subcases run. If I understand correctly, the first update should succeed because it's just putting the interface in the default NI. The second update may fail depending on vendor because the Update does not attempt to remove the interface from the previously assigned NI.

My general suggestion is to avoid order dependence, but doing this test in the order independent way will be more tedious. For the sake of clarity, let's add a comment about the order dependency.

Reply:
This is definitely a bug in the test, let me fix it. I think making the test more robust to either clean up after itself, or do a replace such that the configuration is being declaratively set is definitely required.

Reply:
Made the test clean up after itself.